### PR TITLE
configure: refactor datatype configuration

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Current issues --
* lack of systematic approach, accumulation of MPI-1, MPI-2, MPI-3 patches
* mixing system detection and design logic
* lack of refactoring

The revamp --
* separate feature detection and code logic
     * check language type sizes
     * amend options, e.g. setting disable type sizes to 0
     * generate the datatype values
     * wholistic approach to all bindings -- they need be consistent anyway
* refactor with shell functions and autoconf macros

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

File: confdb/aclocal_datatype.m4 [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+dnl PAC_SET_MPI_TYPE(hexidx, name, len, [multiplier])
+dnl Defines datatype hex handle
+dnl     e.g. PAC_SET_MPI_TYPE(01, MPI_CHAR, 1) -> MPI_CHAR=0x4c000101
+dnl Multiplier can be used to define double types
+dnl     e.g. PAC_SET_MPI_TYPE(48, MPI_COMPLEX, $pac_cv_f77_sizeof_real, 2)
+dnl
+AC_DEFUN([PAC_SET_MPI_TYPE], [
+    if test $3 = 0 ; then
+        $2=MPI_DATATYPE_NULL
+        F77_$2=0
```
Comment (raffenet) [inline confdb/aclocal_datatype.m4:94]: Currently we set `MPI_DATAYPE_NULL` for unavailable types in `mpif.h`, so this change is unexpected.

File: confdb/aclocal_datatype.m4 [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        fi
+    done
+    pac_retval="unavailable"
+}
+
+# return (via $pac_retval) the C type matching integer of size
+get_c_bool_type() {
+    len=$[]1
+    pac_retval=
+    for c_type in _BOOL unsigned_char unsigned_short unsigned_int unsigned_long unsigned_long_long ; do
```
[the remark this answers is not in the document]
Comment (hzhou) [reply confdb/aclocal_datatype.m4]: Yes! 👍 

Fixed.

File: confdb/aclocal_datatype.m4 (hunk 1 above)
Comment (hzhou) [reply confdb/aclocal_datatype.m4:94]: I misread the original code. Thanks for pointing out. Now fixed.

**Comment by hzhou** [comment]
test:mpich/ch3/tcp

**Review by raffenet — APPROVED** [review]
I guess we should pull the trigger. I don't have per-commit ABI testing setup yet, so we should check the nightly report and see if this tripped anything.

**Pull request merged**